File: SourceAFIS/Extraction/Templates/MinutiaCollector.cs
using System;
using System.Collections.Generic;
using System.Text;
using SourceAFIS.General;
using SourceAFIS.Meta;
using SourceAFIS.Extraction.Model;

namespace SourceAFIS.Extraction.Templates
{
    public sealed class MinutiaCollector
    {
        [DpiAdjusted(Min = 2)]
        [Parameter(Lower = 3, Upper = 50)]
        public int DirectionSegmentLength = 29;
        [DpiAdjusted(Min = 0)]
        [Parameter(Lower = 0, Upper = 20)]
        public int DirectionSegmentSkip = 3;
        [DpiAdjusted]
        [Parameter(Upper = 4000)]
        public int DpiScaling = 500;

        public DetailLogger.Hook Logger = DetailLogger.Null;

        byte ComputeDirection(SkeletonBuilder.Ridge ridge)
        {
            int first = DirectionSegmentSkip;
            int last = DirectionSegmentSkip + DirectionSegmentLength - 1;
            
            if (last >= ridge.Points.Count)
            {
                int shift = last - ridge.Points.Count + 1;
                last -= shift;
                first -= shift;
            }
            if (first < 0)
                first = 0;

            return Angle.AtanB(ridge.Points[first], ridge.Points[last]);
        }

        public void Collect(SkeletonBuilder skeleton, TemplateBuilder.MinutiaType type, TemplateBuilder template)
        {
            float dpiFactor = 500 / (float)DpiScaling;
            foreach (SkeletonBuilder.Minutia skeletonMinutia in skeleton.Minutiae)
            {
                if (skeletonMinutia.Valid && skeletonMinutia.Ridges.Count == 1)
                {
                    TemplateBuilder.Minutia templateMinutia = new TemplateBuilder.Minutia();
                    templateMinutia.Type = type;
                    templateMinutia.Position = Calc.Round(Calc.Multiply(dpiFactor, skeletonMinutia.Position));
                    templateMinutia.Direction = ComputeDirection(skeletonMinutia.Ridges[0]);
                    template.Minutiae.Add(templateMinutia);
                }
            }
            Logger.Log(template);
        }
    }
}

File: java/sourceafis/src/sourceafis/extraction/minutiae/MinutiaCollector.java
/**
 * @author Veaceslav Dubenco
 * @since 19.10.2012
 */
package sourceafis.extraction.minutiae;

import sourceafis.extraction.model.Ridge;
import sourceafis.extraction.model.SkeletonBuilder;
import sourceafis.extraction.model.SkeletonBuilderMinutia;
import sourceafis.general.Angle;
import sourceafis.general.DetailLogger;
import sourceafis.meta.DpiAdjusted;
import sourceafis.meta.Parameter;
import sourceafis.templates.Minutia;
import sourceafis.templates.MinutiaType;
import sourceafis.templates.TemplateBuilder;

/**
 * 
 */
public final class MinutiaCollector {
	@DpiAdjusted(min = 2)
	@Parameter(lower = 3, upper = 50)
	public int DirectionSegmentLength = 21;
	@DpiAdjusted(min = 0)
	@Parameter(lower = 0, upper = 20)
	public int DirectionSegmentSkip = 1;

	public DetailLogger.Hook Logger = DetailLogger.off;

	byte ComputeDirection(Ridge ridge) {
		int first = DirectionSegmentSkip;
		int last = DirectionSegmentSkip + DirectionSegmentLength - 1;

		if (last >= ridge.getPoints().size()) {
			int shift = last - ridge.getPoints().size() + 1;
			last -= shift;
			first -= shift;
		}
		if (first < 0)
			first = 0;

		return Angle.AtanB(ridge.getPoints().get(first),
				ridge.getPoints().get(last));
	}

	public void Collect(SkeletonBuilder skeleton, MinutiaType type,
			TemplateBuilder template) {
		for (SkeletonBuilderMinutia skeletonMinutia : skeleton.getMinutiae()) {
			if (skeletonMinutia.Valid
					&& skeletonMinutia.getRidges().size() == 1) {
				Minutia templateMinutia = new Minutia();
				templateMinutia.Type = type;
				templateMinutia.Position = skeletonMinutia.getPosition();
				templateMinutia.Direction = ComputeDirection(skeletonMinutia
						.getRidges().get(0));
				template.minutiae.add(templateMinutia);
			}
		}
		Logger.log(template);
	}
}
